### PR TITLE
feat(query): support adaptive filter reorder

### LINKED ISSUE #12753: Feature: support filters reordering
**Summary**

Some filters may be more effective than others, so we should make it possible to reorder filters.

More selectivity filters will be evaluated asap.

### PR DESCRIPTION
**Pull request by Dousir9**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

During the execution of the filter, for SelectExprs in AND or OR, we can dynamically adjust the execution order of these SelectExpr based on cost to achieve better performance.

## Test Model
To verify the effectiveness of the filter reorder algorithm, we established a model to simulate the execution costs of the filter. The test results show that the algorithm can achieve nearly optimal filter order, as the number of SelectExpr increases, the performance improvement becomes more significant compared to the original filter order.

**Code:** https://github.com/Dousir9/experimental-database-work/blob/main/adaptive_filter.py

**Results:**
截屏2024-01-10 18 58 54

## Test on TPC-H Q19:
截屏2024-01-10 17 33 29

Close #12753

## Tests

- [ ] Unit Test
- [ ] Logic Test
- [ ] Benchmark Test
- [x] No Test  - Covered by existing test.

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [x] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/14295)

### CODE REVIEW

File: src/query/expression/src/filter/select_expr_permutation.rs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                // Reset runtime_sum and runtime_count.
+                self.runtime_sum = runtime;
+                self.runtime_count = 1;
+                // This is a better swap, we keep it at least once.
+                self.keep_better = true;
+            }
+
+            self.observe = false;
+        } else {
+            let random_number = rand::thread_rng().gen_range(0..self.random_number_border);
```
Comment (sundy-li) [inline src/query/expression/src/filter/select_expr_permutation.rs:83]: `rand::thread_rng()` could be a field of `FilterPermutation`

Comment (Dousir9) [reply src/query/expression/src/filter/select_expr_permutation.rs:83]: `FilterPermutation` is a member of `NativeDeserializeDataTransform` and `rand::thread_rng()` does not implement `Send` Trait.

**Review by sundy-li — APPROVED** [review]

File: src/query/expression/src/filter/select_expr_permutation.rs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use rand::Rng;
+
+#[derive(Clone, Debug)]
+pub struct FilterPermutation {
```
Comment (xudong963) [inline src/query/expression/src/filter/select_expr_permutation.rs:18]: Better to add some notes for fields

File: src/query/expression/src/filter/select_expr_permutation.rs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        }
+    }
+
+    pub fn add_statistics(&mut self, runtime: u64) {
+        if !self.can_reorder {
+            return;
+        }
+
+        // Whether there is a swap to be observed.
+        if self.observe {
```
Comment (xudong963) [inline src/query/expression/src/filter/select_expr_permutation.rs:68]: Do we need to add a `iteration_count`?

Comment (Dousir9) [reply src/query/expression/src/filter/select_expr_permutation.rs:68]: No, I have made some improvements and tested them in our model, the result show that our algorithm is better.

**Review by xudong963 — APPROVED** [review]

**Review by BugenZhao — APPROVED** [review]

**Pull request merged**